docs/architecture/adr-068-reverse-sync.md

```diff
@@ -0,0 +1,97 @@
+
+- 20 April 2021: Initial Draft
+
+## Status
+
+Proposed
+
+## Context
+
+The advent of state sync and block pruning gave rise to the opportunity for full nodes to participate in consensus without needing complete block history. This also introduced a problem with respect to evidence handling. Nodes that didn't have all the blocks within the evidence age were incapable of validating evidence, thus halting if that evidence was committed on chain. 
+
+[RFC005](https://github.com/tendermint/spec/blob/master/rfc/005-reverse-sync.md) was published in response to this problem and modified the spec to add a minimum block history invariant. This predominantly sought to extend state sync so that it was capable of fetching and storing the `Header`, `Commit` and `ValidatorSet` (essentially a `LightBlock`) of the last `n` heights, where `n` was calculated based from the evidence age.
+
+This ADR sets out to describe the design of this state sync extension as well as modifications to the light client provider and the merging of tm store.
+
+## Decision
+
+The state sync reactor will be extended by introducing 2 new P2P messages (and a new channel). 
+
+```protobuf
+message LightBlockRequest {
+  uint64 height = 1;
+}
+
+message LightBlockResponse {
+  tendermint.types.LightBlock light_block = 1; 
+}
+```
+
+This will be used by the "reverse sync" protocol that will fetch, verify and store prior light blocks such that the node can safely participate in consensus.
+
+Furthermore this allows for a new light client provider which offers the ability for the `StateProvider` to use the underlying P2P stack instead of RPC.
+
+## Detailed Design
+
+This section will focus first on the reverse sync (here we call it `backfill`) mechanism as a standalone protocol and then look to decribe how it integrates within the state sync reactor and how we define the new p2p light client provider.
+
+```go
+// Backfill fetches verifies and stores necessary history
+// to particapate in consensus / validate evidence
+func (r *Reactor) backfill(state State) error {}
+```
+
+`State` is used to work out how far to go back, namely we need all light blocks that have:
+- a height: `h >= state.LastBlockHeight - state.ConsensusParams.Evidence.MaxAgeNumBlocks`
+- a time: `t >= state.LastBlockTime - state.ConsensusParams.Evidence.MaxAgeDuration`
+
+Reverse Sync relies on two components: A `Dispatcher` and a `BlockQueue`. The `Dispatcher` is a pattern taken from a similar [PR](https://github.com/tendermint/tendermint/pull/4508). It is wired to the `LightBlockChannel` and allows for concurrent light block requests by shifting through a linked list of peers. This abstraction has the nice quality that it can also be used as an array of light providers for a P2P based light client.
+
+The `BlockQueue` is a data structure that allows for multiple workers to fetch light blocks, serializing them for the main thread which picks them off the end of the queue, verifies the hashes and persists them.
```

Comment:
not quite. Here's what I had prototyped:
```
type blockQueue struct {
	mtx sync.Mutex

	// cursors to keep track of which heights need to be fetched and verified
	fetchHeight  int64
	verifyHeight int64

	// termination conditions
	stopHeight int64
	stopTime   time.Time

	// track failed heights (ordered from highest to lowest) so we know what
	// blocks to try fetch again
	failed  []int64
	last    *lightBlockResponse
	pending map[int64]lightBlockResponse
	waiters map[int64]chan lightBlockResponse
}
```
we can't just have a channel because we need to make sure that the verifier reads off the queue in the right order i.e. block 100 then 99 then 98. It's actually largely inspired from the `chunkQueue` in state sync if you're familiar with that.